### ISSUE TITLE
GPU performance range evaluation

### ISSUE THREAD

**Issue by enerc** [tevador/RandomX#24]
I made a quick and dirty test to see how this algo would work with an amd GPU.
This is how I did it:
- I created a 4 GB dataset and filled it with the CPU with random data,
- created one 2MB scratchpad for each GPU threads and fill them with random data on the CPU
- divided the 32 KB of the local data store (LDS) as 2 L1 caches
- used workgroup size 32, with only one thread working out of 16. [on amd the wave size is 16 and has its own Program Counter. ] With a vega56, the card is configured to behave as a 112 processor cores.
- as an alternative, I used workgroup size of 16 with one active thread, one 16 KB L1 and had same results
- created the RandomX sequence as a program with init, 2048 loops of 256 instructions with specified weight and L1/L2 proportions (around 50% each), the dataset mixing with cacheline prefetch, and writing registers to scratchpad.

Before giving the results, some considerations:
- some FPU functions are wrongly implemented (like changing the rounding mode, fscale)
- the compiler might have optimized the sequence of instruction, though I took care of taking the registers random
- all threads execute the same program: but as explained above each wave has it's own Program Counter that can be set with an assembly instruction. GPU code can be created on the CPU, 
DMA transfered (~90 MB/s required), and then executed by flushing the GPU Instruction cache and jumping. BTW, some closed source miners are using this method to run optimized kernels.
The instruction cache is 16 KB shared with 4 CU (32 KB on Vega?), so we have 16/4/2 -> 2 KB of instruction cache available per active thread. That will be short to store the 1900 instructions required inside the main loop (but will be helped with the L2 cache and the fact that 1/4th of the waves are doing something).
- final hash not implemented, but should be very fast on a GPU.

Now the results:
- Vega56 at 1630Mhz, about 9700 h/s
- Vega56 at 1138Mhz, about 6800 h/s
- Fury R9 1050MHz, 5600 h/s

I have seen those things:
- dataset loading has no impact when properly scheduled. This is expected since GPU ram is optimized for consecutive data loading
- the bandwidth for L2/L3 access it about 2.6 GB/s (18 L2/L3 access per loop x 2048 * 8(bytes) x 9700 h/s). This is confirmed by the fact that h/s scales with GPU core frequency.

I checked the consistency of the number of executed GPU instructions:
- 9700 h/s
- times 2048
- times around 1900 ASM instructions for 256 vm instructions (might be lower with hand made asm generation like for x86/arm)
- times 4 clocks per cycle
- divided by 1.6 Ghz
- divided by 112 cores
This gives .78 which sounds ok given double FP add some clocks

Hope I didn't made a huge mistake, but memory bandwidth and number of executed instructions looks ok.
In a  "real" miner, I only see L1 Instruction cache as a potential limiting factor. Initially I was thinking of L2/L3, but HBM is really fast. L1 instruction cache misses will be filled from the L2 cache which is apparently not overloaded. Vega10 has also improved instruction cache.
Needs more investigations so see how the L1 instruction cache impacts performance.

I can put the code (Vulkan) in my github if you want to have a look at this prototype.

**Comment by tevador** [comment]
Good work! Can you make the source code available?

Some notes:

> * the bandwidth for L2/L3 access it about 2.6 GB/s (18 L2/L3 access per loop x 2048 * 8(bytes) x 9700 h/s)

It's a lot more. There are 2 L3 reads and 2 L3 writes per iteration, each 64 bytes. So this alone is about 20 GB/s read + 20 GB/s write (9700 * 16384 * 128).
Then there are additional L2 reads and writes. On average, there are 39 8-byte reads per program and 16 8-byte writes, 25% of which is L2. This gives additional ~17 GB/s of bandwidth at 9700 H/s
So total bandwidth usage of L2+L3 is ~57 GB/s. This is still quite low for a GPU.

> * dataset loading has no impact when properly scheduled. This is expected since GPU ram is optimized for consecutive data loading

Small correction: dataset accesses are random, but it doesn't matter since GPUs will be compute-bound, so the latency will be hidden.

> * all threads execute the same program: but as explained above each wave has it's own Program Counter that can be set with an assembly instruction. GPU code can be created on the CPU,

This may the biggest issue. Can a CPU compile 9700*8 = 77600 kernels per second?

**Comment by hyc** [comment]
Isn't it only 9700 kernels/second?

**Comment by hyc** [comment]
> around 1900 ASM instructions for 256 vm instructions (might be lower with hand made asm generation like for x86/arm)

Can we even find docs for a GPU's actual instruction set? And would we bother - there's a ton more GPU variants and corresponding ISA variants than there are CPUs.

**Comment by tevador** [comment]
@hyc
> Isn't it only 9700 kernels/second?

There are 8 different programs per hash.

**Comment by tevador** [comment]
> 77600 kernels per second

And this assumes only 1 GPU per computer. With a 6 GPU rig, it becomes ~460 000 kernels per second. I'm pretty sure that the standard opencl/vulkan compiler is too slow for this, so you would need a custom JIT compiler for each ISA. Considering how many different GPU ISAs are there, it might take some time.

GPU rigs might need a beefy 8 core CPU.

**Comment by SChernykh** [comment]
@enerc Can you clarify that each hash calculated used different instruction pointer to simulate different random programs as it would be in real miner? Does it generate GCN instructions on the fly during hash calculation or it's all fixed? Real miner will have to generate 8 programs on the fly during each hash calculation. Self-modifying code can make these numbers much worse.

**Comment by enerc** [comment]
I missed the 8 different programs per hash, somewhere. Attached the code. The kernel is is in poc.comp.
[prototype.zip](https://github.com/tevador/RandomX/files/2877304/prototype.zip)

**Comment by tevador** [comment]
@enerc Aha, so Vega 56 could theoretically do ~1200 H/s (9700/8). @hyc was actually right about 9700 kernels per second.

**Comment by enerc** [comment]
@SChernykh No it's a basic GPU program doing the same thing but offline compiled. The purpose was to measure memory bandwidth and possible parallelism.
@tevador So a RandomX loop is 8 loops of 2048 loops of 256 vm, right? In that case, yes the theoretical is 1200 H/s.

**Comment by SChernykh** [comment]
@enerc Yes, I looked in the code and it's fixed sequence of instructions. And there is no divergence, all workers in wavefront execute exactly the same single program - this won't be the case in real miner. You should prepare 16 different programs and enclose them in switch-case to execute 1 program per 1 worker in wavefront.

**Comment by enerc** [comment]
@SChernykh Divergence is not an issue when you run one kernel per wave. Obviously we can't use the 64 parallel lines (4 x 16 on 4 clock cycles). That's why we can't run more than 2 active thread per compute unit. This hurts GPU a lot. A ran it on 112 threads out of the 3584 possible ones. No way to run more threads in //

**Comment by SChernykh** [comment]
But why 112 threads? Vega 56 has only 56 compute units and it can't run 2 threads in parallel because they will diverge. Or do you use 2 threads per 1 random program to do FP instructions?

**Comment by enerc** [comment]
Each compute unit has 4 SIMD, but only 2 can run at the same time. So you can run 2 // threads on a compute unit. That's why in OpenCL LDS is limited to 32KB as there are 64KB on a CU. 2 SIMD running at the same time with 32KB each. I initially thought this was a driver limitation, but it's a HW limitation in fact.

**Comment by SChernykh** [comment]
So it's 1200 h/s for Vega 56 in the end? It's slower than Intel i5-2500K (1350 h/s) according to RandomX tests. And self-modifying code can slow it down even more. But at least GPU speed is not totally destroyed as I thought before.

**Comment by enerc** [comment]
50% of the generated asm instructions are for mulh/smulh/div. If called instead of been inlined that would help a lot on the instruction cache issue.
I will try an other POC with rocm assembly chain tool to see how it goes in a more 'miner like' situation.
Someone to give a short try on GDDR5 with RX580?

**Comment by tevador** [comment]
@enerc Any ideas how nVidia cards might fare? Volta and Turing reportedly have [32 program counters per SM](https://medium.com/@farkhor/per-thread-program-counters-a-tale-of-two-registers-f2061949baf2), so they might be able to run more parallel threads than AMD cards (although they might still be limited by caches).

**Comment by enerc** [comment]
@tevador The Nvidia per thread program counter will allow to run thousands of concurrent RandomX threads. That's a big win over AMD here.
For AMD, I think I could break the 2 SIMDs per CU in assembly, thus having all 4 SIMDs running with 16 kB each.
Let's say we can run 200 threads and we reach the icache size limit at effective 1/4th of GPU speed (because of latencies). L2 would have to provide 200*(5 bytes per instruction on average)*400MHz thus 400 GB/s.
With 12 ports to the L2 cache in Turing, apparently 8 memory controllers per port (from Nvidia drawings), and (I think) at least 64 bits wide bus, I would expect 768 GB/s per GHz of GPU clock. Make sense, since L2 cache is likely to be twice as fast as GDDR6/HBM memory.
The memory used in L2 for icache would be in the order of 1 MB with optimized kernels, so much less than the 6 MB of Turing, and 2 MB of Vega.
I will try to create 100+ stupid 4KB kernels, a jump table and see how it goes.

**Comment by enerc** [comment]
Will open a new issue when I get results.
